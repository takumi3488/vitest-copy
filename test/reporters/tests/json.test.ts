import { resolve } from 'pathe'
import { describe, expect, it } from 'vitest'

import { runVitest } from '../../test-utils'

describe('json reporter', async () => {
  const root = resolve(__dirname, '..', 'fixtures')
  const projectRoot = resolve(__dirname, '..', '..', '..')

  it('generates correct report', async () => {
    const { stdout } = await runVitest({
      reporters: 'json',
      root,
      includeTaskLocation: true,
    }, ['json-fail'])

    const data = JSON.parse(stdout)

    expect(data.testResults).toHaveLength(2)

    const failedImport = data.testResults.find((r: any) => r.name.includes('json-fail-import.test'))!
    const failedTest = data.testResults.find((r: any) => r.name.includes('json-fail.test'))!

    expect(failedTest.assertionResults).toHaveLength(1)
    expect(failedImport.assertionResults).toHaveLength(0)

    expect(failedTest.status).toBe('failed')
    expect(failedImport.status).toBe('failed')

    const result = failedTest.assertionResults[0]
    delete result.duration
    const rootRegexp = new RegExp(projectRoot, 'g')
    result.failureMessages = result.failureMessages
      .map((m: string) => {
        const errorStack = m.split('\n').slice(0, 2).join('\n')
        return errorStack.replace(/\\/g, '/').replace(rootRegexp, '<root>')
      })
    expect(result).toMatchSnapshot()
  }, 40000)

  it.each([
    ['passed', 'all-passing-or-skipped'],
    ['passed', 'all-skipped'],
    ['failed', 'some-failing'],
  ])('resolves to "%s" status for test file "%s"', async (expected, file) => {
    const { stdout } = await runVitest({ reporters: 'json', root }, [file])

    const data = JSON.parse(stdout)

    expect(data.testResults).toHaveLength(1)
    expect(data.testResults[0].status).toBe(expected)
  }, 40000)
})
