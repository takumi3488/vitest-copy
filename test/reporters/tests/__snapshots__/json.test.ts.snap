// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`json reporter > generates correct report 1`] = `
{
  "ancestorTitles": [],
  "failureMessages": [
    "AssertionError: expected 2 to deeply equal 1
    at <root>/test/reporters/fixtures/json-fail.test.ts:8:13",
  ],
  "fullName": "should fail",
  "location": {
    "column": 1,
    "line": 5,
  },
  "meta": {},
  "status": "failed",
  "title": "should fail",
}
`;
